### DISCUSSION TITLE
What happens if the server sends rop3 that the client does not support?

### DISCUSSION THREAD

**Discussion by RyanCode** [FreeRDP/FreeRDP#8969] [Q&A] [answered]
I am studying the source code of X11 client and I have found that freerdp implements all rop3 and rop2 operations (various AND or NOT or OR operations) in software gdi, but why is it not supported for all rop3 operations in X11 or other clients? Will the server not send unsupported rop3 opcodes? I did not find any specific negotiation restrictions in the document to support certain rop3 operations. I look forward to your reply

**Comment by akallabeth** [comment]
you happen to use windows NT 4.0 or similar old software ;)

joking aside, the accelerated `ROP` path was given up a long time ago as the software implementation provides support for all and they are mostly unused nowadays as most data is just copying bitmaps, aside from that drawing path being a fallback as there is the remote FX (windows7/server2008) and `GFX` (windows 8 and newer) paths

the `X11` client is maybe not the best example for learning to use the library as there is some cruft in there and has the most complex implementation (it is the oldest client and has been going through a lot of transformation)

1. the basic rule is, all you need is to copy image data in the `EndPaint` callback as there the drawing to the internal framebuffer is complete.
2. advanced usage is overriding some of the software GDI  callbacks with optimized versions (e.g. use some `GPU` buffers or similar, what is being done with the `X11` client)

that said, step 2 should be long after whatever you try to do is already working ;)

**Reply by RyanCode** [reply, marked as answer]
got it,and I am trying to understand its details, when will it proactively call EndPaint？
